File: FinTracker.Client/src/services/CategoryService.ts
import Category from "../types/Category";
import CategoryReferenceCounts from "../types/CategoryReferenceCounts";
import CategoryViewModel from "../types/CategoryViewModel";
import BaseService from "./BaseService";

class CategoryService extends BaseService {
    constructor() {
        super("/Categories");
    }

    getCategories(): Promise<Category[]> {
        return this.get<Category[]>("");
    }

    getCategoryTransactionCounts(): Promise<CategoryReferenceCounts[]> {
        return this.get<CategoryReferenceCounts[]>("/WithCounts");
    }

    createCategory(model: CategoryViewModel): Promise<Category> {
        return this.post<Category>("", model);
    }

    putCategory(model: CategoryViewModel): Promise<Category> {
        const { id, ...values } = model;
        return this.put<Category>(`/${id}`, values);
    }

    deleteCategory(id: number): Promise<void> {
        return this.delete(`/${id}`);
    }
}

export default new CategoryService();

File: FinTracker.Api/Services/CategoryService.cs
using FinTracker.Api.Common;
using FinTracker.Api.Models;
using FinTracker.Services.Data.Entities;
using Microsoft.EntityFrameworkCore;

namespace FinTracker.Api.Services
{
    public class CategoryService : BaseEntityService<TblCategory>
    {
        public TblCategory? GetCategory(int id)
        {
            return db.TblCategories.Find(id);
        }

        public TblCategory[] GetCategories()
        {
            return db.TblCategories.OrderBy(e => e.CategoryName).ToArray();
        }

        public VwCategoryReferenceCounts[] GetCategoryTransactionCounts()
        {
            return db.VwCategoryReferenceCounts.OrderBy(e => e.CategoryName).ToArray();
        }

        public TblCategory CreateCategory(CategoryViewModel model) => addEntityAndSave(model);
        public TblCategory PutCategory(int id, CategoryViewModel model) => putEntityAndSave(id, model);
        public void DeleteCategory(int id) => deleteEntityAndSave(id);
    }
}
